nullaway/src/main/java/com/uber/nullaway/handlers/SynchronousCallbackHandler.java

```diff
@@ -0,0 +1,93 @@
+package com.uber.nullaway.handlers;
+
+import static com.uber.nullaway.handlers.CompositeHandler.FALSE_AP_PREDICATE;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.suppliers.Supplier;
+import com.google.errorprone.suppliers.Suppliers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.LambdaExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Type;
+import com.uber.nullaway.LibraryModels.MethodRef;
+import com.uber.nullaway.dataflow.AccessPath;
+import java.util.function.Predicate;
+
+public class SynchronousCallbackHandler extends BaseNoOpHandler {
+
+  /**
+   * Maps method name to full information about the corresponding methods and what parameter is the
+   * relevant callback. We key on method name to quickly eliminate most cases when doing a lookup.
+   */
+  private static final ImmutableMap<String, ImmutableMap<MethodRef, Integer>>
+      METHOD_NAME_TO_SIG_AND_PARAM_INDEX =
+          ImmutableMap.of(
+              "forEach",
+              ImmutableMap.of(
+                  MethodRef.methodRef(
+                      "java.util.Map",
+                      "forEach(java.util.function.BiConsumer<? super K,? super V>)"),
+                  0,
+                  MethodRef.methodRef(
+                      "java.lang.Iterable", "forEach(java.util.function.Consumer<? super T>)"),
+                  0),
+              "removeIf",
+              ImmutableMap.of(
+                  MethodRef.methodRef(
+                      "java.util.Collection", "removeIf(java.util.function.Predicate<? super E>)"),
+                  0));
+
+  private static final Supplier<Type> STREAM_TYPE_SUPPLIER =
+      Suppliers.typeFromString("java.util.stream.Stream");
+
+  @Override
+  public Predicate<AccessPath> getAccessPathPredicateForNestedMethod(
+      TreePath path, VisitorState state) {
+    Tree leafNode = path.getLeaf();
+    Preconditions.checkArgument(
+        leafNode instanceof ClassTree || leafNode instanceof LambdaExpressionTree,
+        "Unexpected leaf type: %s",
+        leafNode.getClass());
```

Comment:
Why do we know the `leafNode` is a `ClassTree` or `LambdaExpressionTree` here? (I assume at some point we are filtering out method references and the like?

Edit: Guess it has to do with where we call [updateEnvironmentMapping](https://github.com/uber/NullAway/blob/719b167ded3f65c4eb454d1909f436e9dd514076/nullaway/src/main/java/com/uber/nullaway/NullAway.java#L457). But by the time we get to [getNullnessInfoBeforeNewContext](https://github.com/uber/NullAway/blob/719b167ded3f65c4eb454d1909f436e9dd514076/nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessAnalysis.java#L220) that's no longer very clear in our docs/checks.

Reply:
Good point, tweaked the docs and method name in 34bb9f3909216a4b0109a21d3fa1066bae4eaad8